### PR TITLE
Bug/1372 open template with large datastore

### LINKED ISSUE #1372: Open job as template icw Oracle makes DC hang
Originally submitted by @jacques-hi:

When a job is opened as a template you are offered to choose a different data store to map your job to.
If you select an Oracle datastore, DC will hang.

### PR DESCRIPTION
**Pull request by jhorcicka**
Fixes #1372 

In "Open as a template" dialog when you selected a "large" datastore it took a while to load and the GUI looked frozen. Now there is a loading icon that indicates "something is happening". 

loading-icon

### CODE REVIEW

**Comment by arjansh** [comment]
Code seems to be working as expected, however it seems like the code changes to the SourceColumnComboBox class have no functional impact and are only code styling improvements.

Also the changes to the OpenAnalysisJobAsTemplateDialog are quite extensive and the code is quite complex so it is hard to understand what exactly is going on. Maybe this can be refactored a bit so it becomes more readable (and maybe some unit tests can be added).

**Comment by jhorcicka** [comment]
I refactored OpenAnalysisJobAsTemplateDialog class. @arjansh, can you take a look at it again, please? Perhaps, it would be better to view only new code to understand it. The diff here at github is quite large since I basically moved everything around. :-)

File: desktop/ui/src/main/java/org/datacleaner/windows/OpenAnalysisJobAsTemplateDialog.java [hunk 1]
```diff
@@ -97,163 +101,177 @@ public OpenAnalysisJobAsTemplateDialog(WindowContext windowContext, DataCleanerC
...
-                }
-            });
+        return sourceColumnComboBoxes;
+    }
 
-            if (!_sourceColumnComboBoxes.containsKey(tablePath)) {
-                _sourceColumnComboBoxes.put(tablePath, new ArrayList<SourceColumnComboBox>());
-            }
+    private String getTablePath(String columnPath) {
+        int columnDelim = columnPath.lastIndexOf('.');
```
Comment (arjansh) [inline desktop/ui/src/main/java/org/datacleaner/windows/OpenAnalysisJobAsTemplateDialog.java]: Final.

File: desktop/ui/src/main/java/org/datacleaner/windows/OpenAnalysisJobAsTemplateDialog.java [hunk 2]
```diff
@@ -97,163 +101,177 @@ public OpenAnalysisJobAsTemplateDialog(WindowContext windowContext, DataCleanerC
...
-        _datastoreCombobox = new JComboBox<String>(comboBoxModel);
-        _datastoreCombobox.setEditable(false);
-        _datastoreCombobox.addActionListener(new ActionListener() {
-            @Override
-            public void actionPerformed(ActionEvent e) {
-                final String datastoreName = (String) _datastoreCombobox.getSelectedItem();
-                _datastore = _datastoreCatalog.getDatastore(datastoreName);
+    private JComboBox createDatastoreCombobox() {
+        final String[] comboBoxModel = CollectionUtils.array(new String[1], _datastoreCatalog.getDatastoreNames());
+        final JComboBox comboBox = new JComboBox<>(comboBoxModel);
```
Comment (arjansh) [inline desktop/ui/src/main/java/org/datacleaner/windows/OpenAnalysisJobAsTemplateDialog.java]: JComboBox is a raw type, preferably use JComboBox<String> instead (also for the return type of the method).

File: desktop/ui/src/main/java/org/datacleaner/windows/OpenAnalysisJobAsTemplateDialog.java [hunk 3]
```diff
@@ -97,163 +101,177 @@ public OpenAnalysisJobAsTemplateDialog(WindowContext windowContext, DataCleanerC
...
+        comboBox.setEditable(false);
+        final JDialog parent = this;
+        comboBox.addActionListener(e -> {
+            try {
+                ComboBoxUpdater comboBoxUpdater = new ComboBoxUpdater(parent);
+                comboBoxUpdater.execute();
+            } catch (Exception exception) {
+                logger.error(exception.getMessage());
```
Comment (arjansh) [inline desktop/ui/src/main/java/org/datacleaner/windows/OpenAnalysisJobAsTemplateDialog.java]: Are you sure it is good enough if you just log an error here, or should this be propagated to the user interface in some manner?

**Comment by arjansh** [comment]
@jhorcicka I added some small inline comments. Also I see that Travis fails.

**Comment by jhorcicka** [comment]
The new test fails. It passes locally so I have to look into it.

**Comment by jhorcicka** [comment]
Travis failed because of headless environment and apparently you can't use some "desktop-ui" functionality on such systems.

**Comment by arjansh** [comment]
LGTM!

**Pull request merged**